File: src/cmds/verify.ts
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { spawnSync } from 'child_process';
import { DB, getMeta } from '../core/local-db';
import { TOK_VERSION, fileExistsSync, readFileIfExists } from '../core/utils';
import { readRegisteredClaudeCommand, probeClaudeHook } from '../core/hook';

interface ToolStatus {
  name: string;
  installed: boolean | 'not-detected';
  version?: string;
  hookPath?: string;
  registered?: boolean;
  hookProbe?: 'pass' | 'fail' | 'skipped';
  mode: 'transparent' | 'instruction' | 'unknown';
  note?: string;
}

const TOK_AWARENESS_FILENAME = 'tok-awareness.md';

export function runVerify(db: DB): string {
  const tools: ToolStatus[] = [];

  // Claude Code — transparent (settings.json command hook)
  const claudeHome = path.join(os.homedir(), '.claude');
  const claudeCmd = readRegisteredClaudeCommand();
  if (claudeCmd) {
    const probe = probeClaudeHook(claudeCmd);
    tools.push({
      name: 'Claude Code',
      installed: true,
      hookPath: claudeCmd,
      registered: true,
      hookProbe: probe.pass ? 'pass' : 'fail',
      mode: 'transparent',
    });
  } else if (fileExistsSync(claudeHome)) {
    tools.push({ name: 'Claude Code', installed: false, mode: 'transparent', note: 'detected but hook not registered — run: tok init --claude' });
  } else {
    tools.push({ name: 'Claude Code', installed: 'not-detected', mode: 'unknown' });
  }

  // Cursor — transparent (script in ~/.cursor/hooks/)
  const cursorHome = path.join(os.homedir(), '.cursor');
  const cursorHook = path.join(cursorHome, 'hooks', 'tok-rewrite.sh');
  const cursorCfg = path.join(cursorHome, 'hooks.json');
  if (fileExistsSync(cursorHook)) {
    tools.push({
      name: 'Cursor',
      installed: true,
      version: readHookVersion(cursorHook),
      hookPath: '~/.cursor/hooks/tok-rewrite.sh',
      registered: isCursorRegistered(cursorCfg),
      hookProbe: probeHook(cursorHook, 'cursor'),
      mode: 'transparent',
    });
  } else if (fileExistsSync(cursorHome)) {
    tools.push({ name: 'Cursor', installed: false, mode: 'transparent', note: 'detected but hook missing — run: tok init --cursor' });
  } else {
    tools.push({ name: 'Cursor', installed: 'not-detected', mode: 'unknown' });
  }

  // Awareness-based (instruction mode) — VS Code Copilot, Gemini, Windsurf, Cline
  pushInstructionStatus(tools, 'Copilot (VS Code)', vscodeAwarenessPath());
  pushInstructionStatus(tools, 'Gemini CLI', path.join(os.homedir(), '.gemini', TOK_AWARENESS_FILENAME));
  pushInstructionStatus(tools, 'Windsurf', path.join(os.homedir(), '.codeium', 'windsurf', TOK_AWARENESS_FILENAME));
  pushInstructionStatus(tools, 'Cline / Roo Code', path.join(os.homedir(), '.cline', TOK_AWARENESS_FILENAME));

  const lines: string[] = ['Hook status:'];
  for (const t of tools) {
    if (t.installed === 'not-detected') {
      lines.push(`  -    ${t.name.padEnd(20)} not detected on this system`);
      continue;
    }
    if (t.installed === false) {
      lines.push(`  FAIL ${t.name.padEnd(20)} ${t.note || 'not installed'}`);
      continue;
    }
    const v = t.version ? ` v${t.version}` : '';
    const modeTag = t.mode === 'transparent' ? '[transparent]' : t.mode === 'instruction' ? '[instruction]' : '';
    lines.push(`  OK   ${t.name.padEnd(20)} ${modeTag}${v}   ${t.hookPath || ''}`);
    if (t.registered === false) {
      lines.push(`         WARN: hook script exists but not registered with the AI tool — re-run tok init`);
    }
    if (t.hookProbe === 'pass') {
      lines.push(`         Probe:   PASS (hook produced expected rewrite)`);
    } else if (t.hookProbe === 'fail') {
      lines.push(`         Probe:   FAIL (hook output does not match expected protocol — run: tok hook-test)`);
    }
  }

  const hookV = getMeta(db, 'hook_version');
  lines.push('');
  if (hookV && hookV !== TOK_VERSION) {
    lines.push(`WARN: hooks recorded as v${hookV} but tok is v${TOK_VERSION}. Run: tok init  to refresh.`);
  } else if (hookV) {
    lines.push(`Hooks are current (v${hookV}).`);
  } else {
    lines.push('No hook version recorded yet — run: tok init');
  }
  return lines.join('\n');
}

function readHookVersion(p: string): string | undefined {
  try {
    const content = fs.readFileSync(p, 'utf8');
    const m = /tok-hook-version:\s*([\w.-]+)/.exec(content);
    return m?.[1];
  } catch {
    return undefined;
  }
}

function isCursorRegistered(cfgPath: string): boolean {
  const raw = readFileIfExists(cfgPath);
  if (!raw) return false;
  try {
    const cfg = JSON.parse(raw) as Record<string, unknown>;
    const list1 = (cfg.preToolUse as Array<Record<string, unknown>> | undefined) || [];
    const hooksObj = cfg.hooks as Record<string, unknown> | undefined;
    const list2 = (hooksObj?.preToolUse as Array<Record<string, unknown>> | undefined) || [];
    return [...list1, ...list2].some((h) => h.id === 'tok-rewrite' || /tok-rewrite\.sh/.test(String(h.command || '')));
  } catch {
    return false;
  }
}

function probeHook(hookPath: string, kind: 'claude' | 'cursor'): 'pass' | 'fail' | 'skipped' {
  try {
    const payload = JSON.stringify({ tool_name: 'Bash', tool_input: { command: 'git status' } });
    const r = spawnSync('bash', [hookPath], { input: payload, encoding: 'utf8' });
    if (r.error || r.status !== 0) return 'fail';
    const stdout = (r.stdout || '').trim();
    if (!stdout) return 'fail';
    let parsed: any;
    try { parsed = JSON.parse(stdout); } catch { return 'fail'; }
    const cmd = kind === 'claude'
      ? parsed?.hookSpecificOutput?.updatedInput?.command
      : parsed?.updated_input?.command;
    return typeof cmd === 'string' && cmd.startsWith('tok ') ? 'pass' : 'fail';
  } catch {
    return 'skipped';
  }
}

function vscodeAwarenessPath(): string {
  const candidates = [
    path.join(process.env.APPDATA || '', 'Code', 'User', TOK_AWARENESS_FILENAME),
    path.join(os.homedir(), 'Library', 'Application Support', 'Code', 'User', TOK_AWARENESS_FILENAME),
    path.join(os.homedir(), '.config', 'Code', 'User', TOK_AWARENESS_FILENAME),
  ];
  return candidates.find(fileExistsSync) || candidates[0];
}

function pushInstructionStatus(tools: ToolStatus[], name: string, mdPath: string): void {
  if (fileExistsSync(mdPath)) {
    tools.push({
      name,
      installed: true,
      version: readHookVersion(mdPath),
      hookPath: tildify(mdPath),
      mode: 'instruction',
    });
  } else {
    tools.push({ name, installed: 'not-detected', mode: 'instruction' });
  }
}

function tildify(p: string): string {
  const home = os.homedir();
  return p.startsWith(home) ? p.replace(home, '~') : p;
}

File: internal/doctor/verify.go
package doctor

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/Kevalgor12/tok-proxy/internal/constants"
	"github.com/Kevalgor12/tok-proxy/internal/hook"
	"github.com/Kevalgor12/tok-proxy/internal/store"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

const awarenessFilename = "tok-awareness.md"

var verifyHookVersionRe = regexp.MustCompile(`tok-hook-version:\s*([\w.-]+)`)

type toolStatus struct {
	name       string
	installed  string // "yes" | "no" | "not-detected"
	version    string
	hookPath   string
	registered *bool
	hookProbe  string // "pass" | "fail" | "skipped" | ""
	mode       string // transparent | instruction | unknown
	note       string
}

// RunVerify reports the hook status of every supported tool, plus whether the recorded hook
// version matches this build.
func RunVerify(s *store.Store) string {
	var tools []toolStatus

	// Claude Code - transparent (settings.json command hook).
	claudeHome := filepath.Join(home(), ".claude")
	if claudeCmd := hook.ReadRegisteredClaudeCommand(); claudeCmd != "" {
		pass, _, _ := hook.ProbeClaudeHook()
		tools = append(tools, toolStatus{
			name: "Claude Code", installed: "yes", hookPath: claudeCmd,
			registered: boolPtr(true), hookProbe: passFail(pass), mode: "transparent",
		})
	} else if util.FileExists(claudeHome) {
		tools = append(tools, toolStatus{name: "Claude Code", installed: "no", mode: "transparent",
			note: "detected but hook not registered - run: tok init --claude"})
	} else {
		tools = append(tools, toolStatus{name: "Claude Code", installed: "not-detected", mode: "unknown"})
	}

	// Cursor - instruction rule (~/.cursor/rules/tok.mdc). Cursor's hooks can only allow/deny/ask,
	// never rewrite a command, so tok can't intercept it transparently.
	tools = append(tools, instructionStatus("Cursor", filepath.Join(home(), ".cursor", "rules", "tok.mdc")))

	// Instruction-mode tools.
	tools = append(tools, instructionStatus("Copilot (VS Code)", vscodeAwarenessPath()))
	tools = append(tools, instructionStatus("Gemini CLI", filepath.Join(home(), ".gemini", awarenessFilename)))
	tools = append(tools, instructionStatus("Cline / Roo Code", filepath.Join(home(), ".cline", awarenessFilename)))
	// Antigravity - transparent (PreToolUse "overwrite" hook in the global customization config).
	tools = append(tools, antigravityStatus(filepath.Join(home(), ".gemini", "config", "hooks.json")))
	// Windsurf - instruction rule inside its shared global-rules file.
	tools = append(tools, ruleStatus("Windsurf", filepath.Join(home(), ".codeium", "windsurf", "memories", "global_rules.md")))

	lines := []string{"Hook status:"}
	for _, t := range tools {
		switch t.installed {
		case "not-detected":
			lines = append(lines, "  -    "+util.Pad(t.name, 20, false)+" not detected on this system")
			continue
		case "no":
			lines = append(lines, "  FAIL "+util.Pad(t.name, 20, false)+" "+firstNonEmpty(t.note, "not installed"))
			continue
		}
		v := ""
		if t.version != "" {
			v = " v" + t.version
		}
		modeTag := ""
		switch t.mode {
		case "transparent":
			modeTag = "[transparent]"
		case "instruction":
			modeTag = "[instruction]"
		}
		lines = append(lines, "  OK   "+util.Pad(t.name, 20, false)+" "+modeTag+v+"   "+t.hookPath)
		if t.registered != nil && !*t.registered {
			lines = append(lines, "         WARN: hook script exists but not registered with the AI tool - re-run tok init")
		}
		switch t.hookProbe {
		case "pass":
			lines = append(lines, "         Probe:   PASS (hook produced expected rewrite)")
		case "fail":
			lines = append(lines, "         Probe:   FAIL (hook output does not match expected protocol - run: tok hook-test)")
		}
	}

	hookV, _ := s.GetMeta("hook_version")
	lines = append(lines, "")
	switch {
	case hookV != "" && hookV != constants.Version:
		lines = append(lines, "WARN: hooks recorded as v"+hookV+" but tok is v"+constants.Version+". Run: tok init  to refresh.")
	case hookV != "":
		lines = append(lines, "Hooks are current (v"+hookV+").")
	default:
		lines = append(lines, "No hook version recorded yet - run: tok init")
	}
	return strings.Join(lines, "\n")
}

func readHookVersion(p string) string {
	c, err := os.ReadFile(p)
	if err != nil {
		return ""
	}
	if m := verifyHookVersionRe.FindStringSubmatch(string(c)); m != nil {
		return m[1]
	}
	return ""
}

func vscodeAwarenessPath() string {
	candidates := []string{
		filepath.Join(os.Getenv("APPDATA"), "Code", "User", awarenessFilename),
		filepath.Join(home(), "Library", "Application Support", "Code", "User", awarenessFilename),
		filepath.Join(home(), ".config", "Code", "User", awarenessFilename),
	}
	for _, c := range candidates {
		if util.FileExists(c) {
			return c
		}
	}
	return candidates[0]
}

func instructionStatus(name, mdPath string) toolStatus {
	if util.FileExists(mdPath) {
		return toolStatus{name: name, installed: "yes", version: readHookVersion(mdPath), hookPath: tildify(mdPath), mode: "instruction"}
	}
	return toolStatus{name: name, installed: "not-detected", mode: "instruction"}
}

// ruleStatus reports whether tok's managed rule block is present in a shared rules file
// (Windsurf) - the file itself may exist holding only the user's own rules.
func ruleStatus(name, rulesPath string) toolStatus {
	if c, ok := util.ReadFileIfExists(rulesPath); ok && strings.Contains(c, "<!-- tok:start -->") {
		return toolStatus{name: name, installed: "yes", hookPath: tildify(rulesPath), mode: "instruction"}
	}
	return toolStatus{name: name, installed: "not-detected", mode: "instruction"}
}

// antigravityStatus reports tok's transparent PreToolUse hook in Antigravity's global
// customization config (~/.gemini/config/hooks.json), keyed under "tok".
func antigravityStatus(hooksPath string) toolStatus {
	if c, ok := util.ReadFileIfExists(hooksPath); ok && strings.Contains(c, "hook antigravity") {
		return toolStatus{name: "Antigravity", installed: "yes", hookPath: tildify(hooksPath),
			registered: boolPtr(true), mode: "transparent"}
	}
	geminiDir := filepath.Join(home(), ".gemini")
	if util.FileExists(filepath.Join(home(), ".antigravity")) || util.FileExists(filepath.Join(home(), ".antigravity-ide")) || util.FileExists(geminiDir) {
		return toolStatus{name: "Antigravity", installed: "no", mode: "transparent",
			note: "detected but hook not registered - run: tok init --antigravity"}
	}
	return toolStatus{name: "Antigravity", installed: "not-detected", mode: "transparent"}
}

func tildify(p string) string {
	h := home()
	if h != "" && strings.HasPrefix(p, h) {
		return "~" + p[len(h):]
	}
	return p
}

func boolPtr(b bool) *bool { return &b }

func passFail(pass bool) string {
	if pass {
		return "pass"
	}
	return "fail"
}

func firstNonEmpty(a, b string) string {
	if a != "" {
		return a
	}
	return b
}
